lib/Frontend/Frontend.cpp

```diff
@@ -520,9 +521,39 @@ bool CompilerInstance::setUpModuleLoaders() {
     this->DefaultSerializedLoader = ISML.get();
     Context->addModuleLoader(std::move(ISML));
   }
-  
+
   Context->addModuleLoader(std::move(clangImporter), /*isClang*/ true);
 
+  // When scanning for dependencies, we must add the scanner loaders in order to handle
+  // ASTContext operations such as canImportModule
+  if (Invocation.getFrontendOptions().RequestedAction ==
+      FrontendOptions::ActionType::ScanDependencies) {
+    auto ModuleCachePath = getModuleCachePathFromClang(Context
+                                                       ->getClangModuleLoader()->getClangInstance());
+    auto &FEOpts = Invocation.getFrontendOptions();
+    ModuleInterfaceLoaderOptions LoaderOpts(FEOpts);
+    InterfaceSubContextDelegateImpl ASTDelegate(Context->SourceMgr, Context->Diags,
+                                                Context->SearchPathOpts, Context->LangOpts,
+                                                LoaderOpts,
+                                                Context->getClangModuleLoader(),
+                                                /*buildModuleCacheDirIfAbsent*/false,
+                                                ModuleCachePath,
+                                                FEOpts.PrebuiltModuleCachePath,
+                                                FEOpts.SerializeModuleInterfaceDependencyHashes,
+                                                FEOpts.shouldTrackSystemDependencies());
+    auto mainModuleName = Context->getIdentifier(FEOpts.ModuleName);
+    std::unique_ptr<ModuleDependencyScanner> MDS =
+      std::make_unique<ModuleDependencyScanner>(*Context, MLM, mainModuleName, ASTDelegate);
```

Comment:
Do we need to add `ModuleDependencyScanner` here too? Their `canImport` calls should be properly answered by other module loaders. This also makes me think whether we should add the same parallelism for `PlaceholderSwiftModuleScanner`. We could define a new loader `PlaceholderSwiftModuleLoader` to be added here instead of reusing the scanner.

Reply:
Good point, I don't think we need `ModuleDependencyScanner` to make this work, just the `Placeholder`. 

I think `PlaceholderSwiftModuleLoader` would not be the right abstraction, because it would not be able to actually "load" anything, and functionally would be identical to the corresponding `Scanner`. But I like the line of thinking of making `canImport` behaving similar to the actually dependency discovery. 

Another consideration is that today the two scanners are themselves instantiated by the other module loaders, and not directly (this PR aside). I wonder if we should instead make them first-class citizens, sort of like `ASTContext` has `getImpl().ModuleLoaders`, we could add `ModuleScanners` and then use a single instance of each scanner for all queries? But I am not sure if there maybe anything that would prevent this from working.

Reply:
> I think `PlaceholderSwiftModuleLoader` would not be the right abstraction, because it would not be able to actually "load" anything, and functionally would be identical to the corresponding `Scanner`. But I like the line of thinking of making `canImport` behaving similar to the actually dependency discovery.

Sounds good! We can use `PlaceholderSwiftModuleScanner` as both a scanner and a "loader".

> Another consideration is that today the two scanners are themselves instantiated by the other module loaders, and not directly (this PR aside). I wonder if we should instead make them first-class citizens, sort of like `ASTContext` has `getImpl().ModuleLoaders`, we could add `ModuleScanners` and then use a single instance of each scanner for all queries? But I am not sure if there maybe anything that would prevent this from working.

I'm not a big fan of promoting these scanners to the `ASTContext` because they are specific to `-scan-dependencies` action. IMO, we should hide them as the implementation detail of this action as much as we could.